File: frontend/lib/models/fuel.dart
class NozzleReadingModel {
  final int id;
  final int dailyLogId;
  final int unitId;
  final String openingReading;
  final String closingReading;
  final String grossSaleLiters;

  NozzleReadingModel({
    required this.id,
    required this.dailyLogId,
    required this.unitId,
    required this.openingReading,
    required this.closingReading,
    required this.grossSaleLiters,
  });

  factory NozzleReadingModel.fromJson(Map<String, dynamic> json) {
    return NozzleReadingModel(
      id: json['id'] as int,
      dailyLogId: json['daily_log_id'] as int,
      unitId: json['unit_id'] as int,
      openingReading: json['opening_reading'].toString(),
      closingReading: json['closing_reading'].toString(),
      grossSaleLiters: json['gross_sale_liters'].toString(),
    );
  }
}

class FuelPurchaseModel {
  final int id;
  final int dailyLogId;
  final int productId;
  final int tankId;
  final String? invoiceNo;
  final String purchaseLiters;
  final String purchaseRate;
  final String saleRate;
  final String rateDiffPerLtr;
  final String rateDiffAmount;

  FuelPurchaseModel({
    required this.id,
    required this.dailyLogId,
    required this.productId,
    required this.tankId,
    this.invoiceNo,
    required this.purchaseLiters,
    required this.purchaseRate,
    required this.saleRate,
    required this.rateDiffPerLtr,
    required this.rateDiffAmount,
  });

  factory FuelPurchaseModel.fromJson(Map<String, dynamic> json) {
    return FuelPurchaseModel(
      id: json['id'] as int,
      dailyLogId: json['daily_log_id'] as int,
      productId: json['product_id'] as int,
      tankId: json['tank_id'] as int,
      invoiceNo: json['invoice_no'] as String?,
      purchaseLiters: json['purchase_liters'].toString(),
      purchaseRate: json['purchase_rate'].toString(),
      saleRate: json['sale_rate'].toString(),
      rateDiffPerLtr: json['rate_diff_per_ltr'].toString(),
      rateDiffAmount: json['rate_diff_amount'].toString(),
    );
  }
}

File: backend/app/schemas/fuel.py
from decimal import Decimal
from typing import Optional, List
from pydantic import Field, model_validator
from backend.app.schemas.common import BaseSchema

class NozzleReadingItem(BaseSchema):
    unit_id: int = Field(..., gt=0)
    opening_reading: Decimal = Field(..., ge=Decimal("0.00"))
    closing_reading: Decimal = Field(..., ge=Decimal("0.00"))

    @model_validator(mode="after")
    def check_closing_ge_opening(self):
        if self.closing_reading < self.opening_reading:
            raise ValueError(
                f"Closing reading ({self.closing_reading}) cannot be less than opening reading ({self.opening_reading}) for unit {self.unit_id}."
            )
        return self

class NozzleReadingCreate(BaseSchema):
    daily_log_id: int = Field(..., gt=0)
    readings: List[NozzleReadingItem]

class NozzleReadingResponse(BaseSchema):
    id: int
    daily_log_id: int
    unit_id: int
    opening_reading: Decimal
    closing_reading: Decimal
    gross_sale_liters: Decimal

class FuelPurchaseCreate(BaseSchema):
    daily_log_id: int = Field(..., gt=0)
    product_id: int = Field(..., gt=0)
    tank_id: int = Field(..., gt=0)
    invoice_no: Optional[str] = Field(None, max_length=50)
    purchase_liters: Decimal = Field(..., gt=Decimal("0.00"))
    purchase_rate: Decimal = Field(..., gt=Decimal("0.0000"))
    sale_rate: Decimal = Field(..., gt=Decimal("0.0000"))
    rate_diff_per_ltr: Decimal = Field(Decimal("0.0000"))

class FuelPurchaseResponse(BaseSchema):
    id: int
    daily_log_id: int
    product_id: int
    tank_id: int
    invoice_no: Optional[str] = None
    purchase_liters: Decimal
    purchase_rate: Decimal
    sale_rate: Decimal
    rate_diff_per_ltr: Decimal
    rate_diff_amount: Decimal
